logstash-core/src/main/java/org/logstash/config/ir/compiler/JavaCodecDelegator.java

```diff
@@ -71,52 +65,40 @@ public JavaCodecDelegator(final Context context, final Codec codec) {
 
             final NamespacedMetric encodeMetric = metric.namespace(ENCODE_KEY);
             encodeMetricIn = encodeMetric.counter(IN_KEY);
-            encodeMetricTime = encodeMetric.counter(MetricKeys.DURATION_IN_MILLIS_KEY.asJavaString());
+            encodeMetricTime = encodeMetric.timer(MetricKeys.DURATION_IN_MILLIS_KEY.asJavaString());
 
             final NamespacedMetric decodeMetric = metric.namespace(DECODE_KEY);
             decodeMetricIn = decodeMetric.counter(IN_KEY);
             decodeMetricOut = decodeMetric.counter(MetricKeys.OUT_KEY.asJavaString());
-            decodeMetricTime = decodeMetric.counter(MetricKeys.DURATION_IN_MILLIS_KEY.asJavaString());
+            decodeMetricTime = decodeMetric.timer(MetricKeys.DURATION_IN_MILLIS_KEY.asJavaString());
         }
     }
 
     @Override
     public void decode(final ByteBuffer buffer, final Consumer<Map<String, Object>> eventConsumer) {
         decodeMetricIn.increment();
 
-        final long start = System.nanoTime();
-
-        codec.decode(buffer, (event) -> {
+        decodeMetricTime.time(() -> codec.decode(buffer, (event) -> {
             decodeMetricOut.increment();
             eventConsumer.accept(event);
-        });
-
-        decodeMetricTime.increment(TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS));
+        }));
     }
 
     @Override
     public void flush(final ByteBuffer buffer, final Consumer<Map<String, Object>> eventConsumer) {
         decodeMetricIn.increment();
 
-        final long start = System.nanoTime();
-
-        codec.flush(buffer, (event) -> {
+        decodeMetricTime.time(() -> codec.flush(buffer, (event) -> {
             decodeMetricOut.increment();
             eventConsumer.accept(event);
-        });
-
-        decodeMetricTime.increment(TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS));
+        }));
     }
 
     @Override
     public void encode(final Event event, final OutputStream out) throws IOException {
         encodeMetricIn.increment();
 
-        final long start = System.nanoTime();
-
-        codec.encode(event, out);
-
-        decodeMetricTime.increment(TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS));
```

Comment:
oh look, a bug. This is likely a part of why the encode metrics weren't getting tracked correctly.